### PR TITLE
Member Management Part2

### LINKED ISSUE #788: Member Management Improvements
The current implementation of the project member management page has some limitations that we know about. Now with the requirement to connect service accounts from different namespaces, these limitations need to be resolved.

Screen Shot 2020-08-26 at 17 53 12

# Planned changes

## Project User Card
- [x]  Change the _trash_ button to _remove_ (as no actual deletion takes place this may causes confusion)
- [x]  Prevent to add service accounts via the add user dialog (prevent names that start with _system:serviceaccount_), the reason for this is that it may confuses users when they add a service account using this dialog and the connected account appears on a different card
- [x]  Show owner role (grayed out, not editable, with tooltip)

## Service Accounts Card
Currently, the name of tis card is a bit misleading as we actually do not show (all) service accounts. We only show service accounts that have a role in the project, more precisely service accounts that have en entry of kind _User_ in the member section of the _project_ resource. This can lead to name clashes even if no service account with the name is listed.

- [x]  Show all service accounts, even those that have no role in the project
- [x]  Possibility to add service accounts without a role
- [x]  Deletion of service accounts need to be confirmed by entering the name
- [x]  Possibility to connect service accounts from different namespaces by using the full qualified name. Those service accounts need to have a role and cannot be deleted, so they also get a _remove_ button
- [x]  Add service accounts to project resource with kind _ServiceAccount_
- [x]  Add possibility to rotate service account secrets
- [x] Show service account details in row
- [x] Support descriptions for Service Accounts (annotation)

## Other Changes
- [x]  Add information about role types to help dialog
- [x]  Rename _Technical Contact_ to owner on Administration page
- [x]  Remove _Technical Contact_ Card

### PR DESCRIPTION
**Pull request by grolu**
**What this PR does / why we need it**:
This PR adds the missing functionality for the member management improvements

**Which issue(s) this PR fixes**:
Fixes #788 

**Special notes for your reviewer**:

**Release note**:

```feature user
Member Management Improvements
- Added feature that allows to rotate secrets of service accounts
- Members are now sortable and shown in table layout
- Added help texts for member roles
- Service Account Members whose Service Account has been deleted, are now flagged in the Dashboard and can be cleaned up
```

### CODE REVIEW

File: backend/lib/services/members/MemberManager.js [hunk 1]
```diff
@@ -97,6 +97,21 @@ class MemberManager {
...
   }
 
+  async deleteSecret (id) {
+    const item = this.subjectList.get(id)
+    if (!item) {
+      return this.subjectList.members
+    }
+
+    if (!item.kind === 'ServiceAccount') {
+      throw new MethodNotAllowed('Member is not a ServiceAccount')
```
Comment (holgerkoser) [inline backend/lib/services/members/MemberManager.js:107]:
```suggestion
      throw new UnprocessableEntity('Member is not a ServiceAccount')
```

File: backend/lib/routes/members.js [hunk 2]
```diff
@@ -66,3 +66,15 @@ router.route('/:name')
       next(err)
     }
   })
+
+router.route('/:name/secret')
+  .delete(async (req, res, next) => {
+    try {
+      const user = req.user
+      const namespace = req.params.namespace
+      const name = req.params.name
```
Comment (holgerkoser) [inline backend/lib/routes/members.js:75]:
```suggestion
      const { namespace, name } = req.params
```

File: backend/lib/routes/members.js [hunk 3]
```diff
@@ -66,3 +66,15 @@ router.route('/:name')
       next(err)
     }
   })
+
+router.route('/:name/secret')
```
Comment (holgerkoser) [inline backend/lib/routes/members.js:70]: there is no route get secret. I would use verb `post` on the member route.

File: backend/lib/services/members/MemberManager.js [hunk 4]
```diff
@@ -97,6 +97,21 @@ class MemberManager {
     return this.subjectList.members
   }
 
+  async deleteSecret (id) {
+    const item = this.subjectList.get(id)
+    if (!item) {
+      return this.subjectList.members
```
Comment (holgerkoser) [inline backend/lib/services/members/MemberManager.js:103]: We do not modify the member list. I would not return a body?

File: backend/lib/services/members/MemberManager.js [hunk 5]
```diff
@@ -97,6 +97,21 @@ class MemberManager {
...
+    }
+
+    if (!item.kind === 'ServiceAccount') {
+      throw new MethodNotAllowed('Member is not a ServiceAccount')
+    }
+
+    await this.deleteServiceAccountSecret(item)
+
+    return this.subjectList.members
```
Comment (holgerkoser) [inline backend/lib/services/members/MemberManager.js:112]: We do not modify the member list. I would not return a body?

File: backend/lib/services/members/MemberManager.js [hunk 6]
```diff
@@ -97,6 +97,21 @@ class MemberManager {
     return this.subjectList.members
   }
 
+  async deleteSecret (id) {
```
Comment (holgerkoser) [inline backend/lib/services/members/MemberManager.js:100]:
```suggestion
  async rotateServiceAccountSecret (id) {
```

File: backend/lib/services/members/MemberManager.js [hunk 7]
```diff
@@ -169,6 +184,21 @@ class MemberManager {
     this.subjectList.delete(item.id)
   }
 
+  async deleteServiceAccountSecret (item) {
+    const { namespace } = Member.parseUsername(item.id)
+    if (namespace !== this.namespace) {
+      return
```
Comment (holgerkoser) [inline backend/lib/services/members/MemberManager.js:190]: This should be an error I think?!

File: backend/lib/services/members/MemberManager.js [hunk 8]
```diff
@@ -169,6 +184,21 @@ class MemberManager {
...
+    }
+
+    const secretName = _
+      .chain(item)
+      .get('extensions.secrets')
+      .head()
```
Comment (holgerkoser) [inline backend/lib/services/members/MemberManager.js:196]: What if we have more than one secret?

File: backend/lib/services/members/MemberManager.js [hunk 9]
```diff
@@ -169,6 +184,21 @@ class MemberManager {
...
+      .get('name')
+      .value()
+    return await this.client.core.secrets.delete(namespace, secretName)
```
Comment (holgerkoser) [inline backend/lib/services/members/MemberManager.js:199]:
```suggestion
    const name = _
      .chain(item)
      .get('extensions.secrets')
      .head()
      .get('name')
      .value()
    return await this.client.core.secrets.delete(namespace, name)
```

File: frontend/src/components/ProjectServiceAccountRow.vue [hunk 10]
```diff
@@ -129,41 +122,11 @@ export default {
...
-    created: {
-      type: String
-    },
-    roles: {
-      type: Array,
-      required: true
-    },
-    roleDisplayNames: {
-      type: Array,
+    item: {
```
Comment (holgerkoser) [inline frontend/src/components/ProjectServiceAccountRow.vue:129]: Why do we change the explicit behavior to an implicit untyped?

File: frontend/src/store/modules/members.js [hunk 11]
```diff
@@ -41,6 +41,11 @@ const actions = {
     const namespace = rootState.namespace
     const res = await deleteMember({ namespace, name })
     commit('RECEIVE', res.data)
+  },
+  async rotateServiceAccountSecret ({ commit, rootState }, name) {
+    const namespace = rootState.namespace
+    const res = await deleteServiceAccountSecret({ namespace, name })
+    commit('RECEIVE', res.data)
```
Comment (holgerkoser) [inline frontend/src/store/modules/members.js:48]: We do not change the list

File: frontend/src/views/Members.vue [hunk 12]
```diff
@@ -43,24 +43,22 @@ SPDX-License-Identifier: Apache-2.0
           Access to resources within your project can be configured by assigning roles.
         </p>
       </v-card-text>
-      <v-list two-line subheader v-else>
-        <template v-for="(user, index) in sortedAndFilteredUserList">
-          <v-divider v-if="index !== 0" inset :key="index"></v-divider>
+      <v-data-table
```
Comment (holgerkoser) [inline frontend/src/views/Members.vue:46]: not sortable, no pagination, no filter ==> <v-simple-table>?

File: frontend/src/views/Members.vue [hunk 13]
```diff
@@ -321,7 +374,9 @@ export default {
     ...mapActions([
       'addMember',
       'deleteMember',
-      'setError'
+      'rotateServiceAccountSecret',
+      'setError',
+      'set'
```
Comment (holgerkoser) [inline frontend/src/views/Members.vue:379]: There is no action `set` on the store!

File: backend/lib/services/members/MemberManager.js (hunk 4 above)
Comment (grolu) [reply backend/lib/services/members/MemberManager.js:103]: You are right. I think I just did it as we always did in the past. I will have another look.

File: backend/lib/services/members/MemberManager.js (hunk 7 above)
Comment (grolu) [reply backend/lib/services/members/MemberManager.js:190]: Yes, we can throw an error here. Also 422 in this case?

File: backend/lib/services/members/MemberManager.js (hunk 8 above)
Comment (grolu) [reply backend/lib/services/members/MemberManager.js:196]: Hmm, well I asked myself the same question. The way we generate kubeconfigs is that we take the first secret so I guessed, deleting the first one will do the job. I don't know how we should handle service accounts with multiple secrets. We could of course delete all, but I don't know if this what we want. WDYT?

File: frontend/src/components/ProjectServiceAccountRow.vue (hunk 10 above)
Comment (grolu) [reply frontend/src/components/ProjectServiceAccountRow.vue:129]: @petersutter refactored the code and added this component for reuse in the terminal feature context. However, he missed to re-use the component in the user context but it is the very same code so re-use here makes 100% sense. However, the name `ServiceAccount` makes no sense if we use it ofr all account types (User and Service Accounts)

File: frontend/src/views/Members.vue (hunk 13 above)
Comment (grolu) [reply frontend/src/views/Members.vue:379]: IDK where this came from. rm

File: frontend/src/store/modules/members.js (hunk 11 above)
Comment (grolu) [reply frontend/src/store/modules/members.js:48]: I removed the commit, you are right ... same as above

File: frontend/src/views/Members.vue (hunk 12 above)
Comment (grolu) [reply frontend/src/views/Members.vue:46]: Filter is in place, sorting can be added if we want but I currently do not see a real use case for it... maybe for the roles but I don't know... I will add pagination and adjust the row spacing and alignment, as we discussed with @petersutter

File: backend/lib/services/members/MemberManager.js (hunk 8 above)
Comment (grolu) [reply backend/lib/services/members/MemberManager.js:194]: we can throw an error if there is more than one secret (422)

File: backend/lib/routes/members.js [hunk 14]
```diff
@@ -56,24 +56,27 @@ router.route('/:name')
...
-      const name = req.params.name
-      res.send(await members.remove({ user, namespace, name }))
+      const { namespace, name } = req.params
+      const { method } = req.body
+
+      if (method === 'rotateSecret') {
+        res.send(await members.rotateSecret({ user, namespace, name }))
+      } else {
+        throw new Error(`${method} not allowed for members`)
+      }
```
Comment (holgerkoser) [inline backend/lib/routes/members.js:69]:
```suggestion
      switch (method) {
        case 'rotateSecret':
          res.send(await members.rotateSecret({ user, namespace, name }))
          break
        default:
          throw new UnprocessableEntity(`${method} not allowed for members`)
      }
```

File: backend/lib/services/members/MemberManager.js [hunk 15]
```diff
@@ -159,7 +157,7 @@ class MemberManager {
   async updateServiceAccount (item, { description }) {
     const { namespace, name } = Member.parseUsername(item.id)
     if (namespace !== this.namespace) {
-      return
+      throw new UnprocessableEntity(`ServiceAccount namespace ${namespace} does not match project namespace this.namespace`)
```
Comment (holgerkoser) [inline backend/lib/services/members/MemberManager.js:160]:
```suggestion
      throw new UnprocessableEntity('It is not possible to modify ServiceAccount from another namespace')
```

File: backend/lib/services/members/MemberManager.js [hunk 16]
```diff
@@ -187,16 +185,20 @@ class MemberManager {
   async deleteServiceAccountSecret (item) {
     const { namespace } = Member.parseUsername(item.id)
     if (namespace !== this.namespace) {
-      return
+      throw new UnprocessableEntity(`ServiceAccount namespace ${namespace} does not match project namespace this.namespace`)
```
Comment (holgerkoser) [inline backend/lib/services/members/MemberManager.js:188]:
```suggestion
      throw new UnprocessableEntity(`ServiceAccount namespace ${namespace} does not match project namespace this.namespace`)
```
```suggestion
   throw new UnprocessableEntity('It is not possible to modify ServiceAccount from another namespace')
```

File: backend/lib/services/members/MemberManager.js [hunk 17]
```diff
@@ -187,16 +185,20 @@ class MemberManager {
...
     }
 
-    const secretName = _
+    if (_.get(item, 'extensions.secrets.length') > 1) {
```
Comment (holgerkoser) [inline backend/lib/services/members/MemberManager.js:191]:
```suggestion
    if (_.size(item, 'extensions.secrets') > 1) {
```

File: backend/lib/services/members/MemberManager.js [hunk 18]
```diff
@@ -187,16 +185,20 @@ class MemberManager {
...
+      throw new UnprocessableEntity(`ServiceAccount ${namespace} has more than one secret`)
```
Comment (holgerkoser) [inline backend/lib/services/members/MemberManager.js:192]: I am not sure if this should be an client error

File: backend/test/services.members.spec.js [hunk 19]
```diff
@@ -393,6 +412,15 @@ describe('services', function () {
         })
       })
 
+      describe('#deleteServiceAccount', function () {
+        it('should delete a serviceaccount', async function () {
+          const id = 'system:serviceaccount:garden-foo:robot-sa'
+          const item = memberManager.subjectList.get(id)
+          await memberManager.deleteServiceAccount(item)
+          expect(client.core.serviceaccounts.delete).to.have.been.calledOnce
```
Comment (holgerkoser) [inline backend/test/services.members.spec.js:420]: also check arguments

File: backend/test/services.members.spec.js [hunk 20]
```diff
@@ -403,11 +431,50 @@ describe('services', function () {
...
+          }
+        })
+      })
+
+      describe('#deleteServiceAccountSecret', function () {
+        it('should delete a serviceaccount secret', async function () {
+          const id = 'system:serviceaccount:garden-foo:robot-sa'
+          const item = memberManager.subjectList.get(id)
+          await memberManager.deleteServiceAccountSecret(item)
+          expect(client.core.secrets.delete).to.have.been.calledOnce
```
Comment (holgerkoser) [inline backend/test/services.members.spec.js:451]: check arguments

File: backend/test/services.members.spec.js [hunk 21]
```diff
@@ -403,11 +431,50 @@ describe('services', function () {
...
       describe('#updateServiceAccount ', function () {
-        it('should not delete a serviceaccount from a different namespace', async function () {
+        it('should not update a serviceaccount from a different namespace', async function () {
+          const id = 'system:serviceaccount:garden-foreign:robot-foreign-namespace'
+          const item = memberManager.subjectList.get(id)
+          try {
+            await memberManager.updateServiceAccount(item, {})
+            expect.fail('should throw an error')
+          } catch (err) {
+            expect(err).to.be.instanceof(UnprocessableEntity)
```
Comment (holgerkoser) [inline backend/test/services.members.spec.js:441]: We have many UnprocessableEntity errors check significant part of message with regex

File: frontend/src/components/ProjectServiceAccountRow.vue [hunk 22]
```diff
@@ -5,121 +5,114 @@ SPDX-License-Identifier: Apache-2.0
...
+
+          <span class="body-2">{{item.username}}</span>
+        </div>
+      </div>
+    </td>
+    <td>
+      <account-avatar :account-name="item.createdBy" :size="16"></account-avatar>
+    </td>
+    <td>
+      <div>
```
Comment (holgerkoser) [inline frontend/src/components/ProjectServiceAccountRow.vue:33]: Why this div

**Comment by petersutter** [comment]
- The edit button of service accounts has the tooltip "Change Service Account Roles", even though you can edit the description there
- Should we use the `EditableText` component do edit the description in place (in the table)?

File: frontend/src/components/ProjectServiceAccountRow.vue (hunk 22 above)
Comment (grolu) [reply frontend/src/components/ProjectServiceAccountRow.vue:33]: switched back to v-list

File: backend/lib/services/members/MemberManager.js [hunk 23]
```diff
@@ -97,6 +97,19 @@ class MemberManager {
     return this.subjectList.members
   }
 
+  async rotateServiceAccountSecret (id) {
+    const item = this.subjectList.get(id)
+    if (!item) {
+      return
+    }
+
+    if (!item.kind === 'ServiceAccount') {
```
Comment (holgerkoser) [inline backend/lib/services/members/MemberManager.js:106]: Very strange kind of boolean algebra
```suggestion
    if (item.kind !== 'ServiceAccount') {
```

Comment (holgerkoser) [reply backend/lib/services/members/MemberManager.js:106]: Don't we have a unit test for this case?

File: backend/lib/services/members/MemberManager.js [hunk 24]
```diff
@@ -162,13 +175,32 @@ class MemberManager {
...
+      throw new UnprocessableEntity('It is not possible to modify a ServiceAccount from another namespace')
+    }
+
+    if (_.size(_.get(item, 'extensions.secrets')) > 1) {
+      throw new UnprocessableEntity(`ServiceAccount ${namespace} has more than one secret`)
+    }
+
+    const name = _
+      .chain(item)
+      .get('extensions.secrets')
```
Comment (holgerkoser) [inline backend/lib/services/members/MemberManager.js:197]:
```suggestion
      .get('extensions.secrets', [])
      .tap(secrets => {
        if (secrets.length > 1) {
          throw new UnprocessableEntity(`ServiceAccount ${namespace} has more than one secret`)
        }
      })
```

File: backend/lib/services/members/MemberManager.js [hunk 25]
```diff
@@ -162,13 +175,32 @@ class MemberManager {
...
 
+  async deleteServiceAccountSecret (item) {
+    const { namespace } = Member.parseUsername(item.id)
+    if (namespace !== this.namespace) {
+      throw new UnprocessableEntity('It is not possible to modify a ServiceAccount from another namespace')
+    }
+
+    if (_.size(_.get(item, 'extensions.secrets')) > 1) {
+      throw new UnprocessableEntity(`ServiceAccount ${namespace} has more than one secret`)
+    }
```
Comment (holgerkoser) [inline backend/lib/services/members/MemberManager.js:193]:
```suggestion
```

File: frontend/src/components/ProjectServiceAccountRow.vue [hunk 26]
```diff
@@ -5,164 +5,134 @@ SPDX-License-Identifier: Apache-2.0
...
+              </v-btn>
+            </template>
+            <span>Change Service Account</span>
+          </v-tooltip>
+        </div>
+        <div v-if="canManageServiceAccountMembers" class="ml-1">
+          <v-tooltip top>
+            <template v-slot:activator="{ on }">
+              <v-btn v-on="on" icon color="red" @click.native.stop="onDelete">
+                <v-icon v-if="isServiceAccountFromCurrentNamespace">mdi-delete</v-icon>
```
Comment (holgerkoser) [inline frontend/src/components/ProjectServiceAccountRow.vue:99]: Why not a computed property?

File: frontend/src/components/ProjectServiceAccountRow.vue [hunk 27]
```diff
@@ -5,164 +5,134 @@ SPDX-License-Identifier: Apache-2.0
...
+                <v-icon v-else>mdi-close</v-icon>
+              </v-btn>
+            </template>
+            <span v-if="isServiceAccountFromCurrentNamespace">Delete Service Account</span>
+            <span v-else>Remove Foreign Service Account from Project</span>
```
Comment (holgerkoser) [inline frontend/src/components/ProjectServiceAccountRow.vue:104]: Why not a computed property?

File: frontend/src/components/ProjectServiceAccountRow.vue [hunk 28]
```diff
@@ -173,41 +143,47 @@ export default {
...
       return namespace
+    },
+    selectedHeaders () {
+      return mapTableHeader(this.headers, 'selected')
     }
   },
   methods: {
     onDownload () {
-      this.$emit('download', this.username)
+      this.$emit('download', this.item.username)
```
Comment (holgerkoser) [inline frontend/src/components/ProjectServiceAccountRow.vue:165]: Always emit the item

File: frontend/src/components/ProjectUserRow.vue [hunk 29]
```diff
@@ -5,109 +5,88 @@ SPDX-License-Identifier: Apache-2.0
...
+    </td>
+    <td width="250px" v-if="selectedHeaders.actions">
+      <div class="d-flex flex-row justify-end">
+        <div v-if="canManageMembers" class="ml-1">
+          <v-tooltip top>
             <template v-slot:activator="{ on }">
-              <v-chip v-on="on" class="mr-3" small :color="notEditable ? 'grey' : 'black'" outlined>
-                {{displayName}}
-              </v-chip>
+              <v-btn v-on="on" icon @click.native.stop="onEdit">
```
Comment (holgerkoser) [inline frontend/src/components/ProjectUserRow.vue:31]:
```suggestion
              <v-btn v-on="on" icon @click.stop="onEdit">
```

File: backend/lib/services/members/MemberManager.js (hunk 23 above)
Comment (grolu) [reply backend/lib/services/members/MemberManager.js:106]: we have a test and if the expression above is wrong the test fails

File: frontend/src/components/ProjectUserRow.vue [hunk 30]
```diff
@@ -5,109 +5,88 @@ SPDX-License-Identifier: Apache-2.0
...
-    ])
+    ]),
+    selectedHeaders () {
+      return mapTableHeader(this.headers, 'selected')
+    }
   },
   methods: {
     onDelete (username) {
-      this.$emit('delete', this.username)
+      this.$emit('delete', this.item.username)
```
Comment (holgerkoser) [inline frontend/src/components/ProjectUserRow.vue:86]: emit item

File: frontend/src/components/TableColumnSelection.vue [hunk 31]
```diff
@@ -0,0 +1,129 @@
...
+    }
+  },
+  data () {
+    return {
+      columnSelectionMenu: false
+    }
+  },
+  methods: {
+    setSelectedColumn (header) {
+      this.$emit('setSelectedColumn', header)
```
Comment (holgerkoser) [inline frontend/src/components/TableColumnSelection.vue:105]: The names of the events should not be the name of method that the listener will have to call. It should be something like `selectionChange` or `reset` or `filterChange`.

File: frontend/src/views/Members.vue [hunk 32]
```diff
@@ -43,24 +48,26 @@ SPDX-License-Identifier: Apache-2.0
...
       </v-card-text>
-      <v-list two-line subheader v-else>
-        <template v-for="(user, index) in sortedAndFilteredUserList">
-          <v-divider v-if="index !== 0" inset :key="index"></v-divider>
+      <v-data-table
+        v-else
+        :headers="visibleUserAccountTableHeaders"
+        :items="userList"
+        :footer-props="{ 'items-per-page-options': [5,10,20] }"
+        :options.sync="userAccountTableOptions"
```
Comment (holgerkoser) [inline frontend/src/views/Members.vue:56]: Here we use a sync property to distribute shared state across components. I think we should only use one possible solution for one type of problem. Why don't we share the state via the store. Even for reuse components we could have a path which directs the component to the correct store.state property. Also the localStorage handling should be done centrally in the store.

File: backend/lib/routes/members.js (hunk 3 above)
Comment (grolu) [reply backend/lib/routes/members.js]: done

File: frontend/src/components/ProjectUserRow.vue (hunk 30 above)
Comment (holgerkoser) [reply frontend/src/components/ProjectUserRow.vue:86]: username parameter always empty ==> delete

File: frontend/src/components/ProjectServiceAccountRow.vue (hunk 26 above)
Comment (grolu) [reply frontend/src/components/ProjectServiceAccountRow.vue:99]: optimized to not use templating conditions

File: frontend/src/components/ProjectServiceAccountRow.vue (hunk 27 above)
Comment (grolu) [reply frontend/src/components/ProjectServiceAccountRow.vue:104]: optimized to not use templating conditions

File: frontend/src/components/ProjectServiceAccountRow.vue (hunk 28 above)
Comment (grolu) [reply frontend/src/components/ProjectServiceAccountRow.vue:165]: done

File: frontend/src/components/ProjectUserRow.vue (hunk 30 above)
Comment (grolu) [reply frontend/src/components/ProjectUserRow.vue:86]: done

File: frontend/src/components/TableColumnSelection.vue (hunk 31 above)
Comment (grolu) [reply frontend/src/components/TableColumnSelection.vue:105]: renamed as suggested

Comment (grolu) [reply frontend/src/components/TableColumnSelection.vue:105]: we agreed to do this in a separate PR, see also https://github.com/gardener/dashboard/issues/905

File: frontend/src/views/Members.vue (hunk 32 above)
Comment (grolu) [reply frontend/src/views/Members.vue:56]: we agreed to do this in a separate PR, see also https://github.com/gardener/dashboard/issues/905

File: backend/lib/services/members/MemberManager.js [hunk 33]
```diff
@@ -162,13 +175,33 @@ class MemberManager {
   async deleteServiceAccount (item) {
     const { namespace, name } = Member.parseUsername(item.id)
     if (namespace !== this.namespace) {
-      return
+      return // do not throw an error as it is allowed to add foreign service accounts
```
Comment (petersutter) [inline backend/lib/services/members/MemberManager.js:178]: is it really allowed to delete a service account from another namespace?
if yes, the comment needs to be adapted anyhow: `add` -> `delete`

File: frontend/src/views/ShootList.vue [hunk 34]
```diff
@@ -26,125 +26,13 @@ SPDX-License-Identifier: Apache-2.0
...
-                      <div v-for="label in hideClustersWithLabels" :key="label">- {{label}}</div>
-                    </v-tooltip>
-                    ticket labels
-                  </v-list-item-title>
-                </v-list-item-content>
-              </v-list-item>
-            </template>
-          </v-list>
-        </v-menu>
+        <table-column-selection
```
Comment (petersutter) [inline frontend/src/views/ShootList.vue:29]: bug: I cannot select any columns under the table options menu

File: frontend/src/components/ProjectServiceAccountRow.vue [hunk 35]
```diff
@@ -5,164 +5,132 @@ SPDX-License-Identifier: Apache-2.0
...
+          </v-tooltip>
+        </div>
+        <div v-if="canManageServiceAccountMembers" class="ml-1">
+          <v-tooltip top>
+            <template v-slot:activator="{ on }">
+              <v-btn v-on="on" icon @click.native.stop="onEdit">
+                <v-icon>mdi-pencil</v-icon>
+              </v-btn>
+            </template>
+            <span>Change Service Account</span>
```
Comment (petersutter) [inline frontend/src/components/ProjectServiceAccountRow.vue:92]: When I change the role of a service account and press `Update` it takes very long until the dialog closes. Did anything change here? At first I thought it didn't react on my button press but then after ~3 seconds the dialog closed

**Review by petersutter — CHANGES_REQUESTED** [review]
there are some bugs that needs to be fixed, and see my other comments

File: frontend/src/views/Members.vue [hunk 36]
```diff
@@ -454,12 +552,23 @@ export default {
         confirmValue: name
       })
     },
-    onEditUser (username, roles) {
+    confirmRotateServiceAccountSecret (name) {
+      const memberName = escape(displayName(name))
+      const messageHtml = `Do you want to rotate the secret of service account <i>${memberName}</i>?<br /> Attention: The current kubeconfig credentials will be revoked`
```
Comment (petersutter) [inline frontend/src/views/Members.vue:557]: change to vue templates

File: backend/lib/services/members/MemberManager.js (hunk 33 above)
Comment (grolu) [reply backend/lib/services/members/MemberManager.js:178]: renamed comment to make it clearer

File: frontend/src/components/ProjectServiceAccountRow.vue (hunk 35 above)
Comment (grolu) [reply frontend/src/components/ProjectServiceAccountRow.vue:92]: I did not change the way this works. Probably some network / server load issue. But there is certainly room for improvement when handling these async actions on the UI, e.g. disabling the button, adding a spinner...
Probably not scope of this PR though

File: frontend/src/views/Members.vue (hunk 36 above)
Comment (grolu) [reply frontend/src/views/Members.vue:557]: done

File: frontend/src/views/ShootList.vue (hunk 34 above)
Comment (grolu) [reply frontend/src/views/ShootList.vue:29]: not reducable

**Review by petersutter — APPROVED** [review]
/lgtm

File: backend/lib/routes/members.js [hunk 37]
```diff
@@ -56,6 +57,23 @@ router.route('/:name')
...
   })
+  .post(async (req, res, next) => {
+    try {
+      const user = req.user
+      const { namespace, name } = req.params
+      const { method } = req.body
+
+      switch (method) {
+        case 'rotateSecret':
+          res.send(await members.rotateSecret({ user, namespace, name }))
```
Comment (holgerkoser) [inline backend/lib/routes/members.js:68]:
```suggestion
          await members.rotateSecret({ user, namespace, name })
          res.end()
```

File: backend/lib/services/members/MemberManager.js [hunk 38]
```diff
@@ -162,13 +175,33 @@ class MemberManager {
...
+      .get('extensions.secrets', [])
+      .tap(secrets => {
+        if (secrets.length > 1) {
+          throw new UnprocessableEntity(`ServiceAccount ${namespace} has more than one secret`)
+        }
+      })
+      .head()
+      .get('name')
+      .value()
+    return await this.client.core.secrets.delete(namespace, name)
```
Comment (holgerkoser) [inline backend/lib/services/members/MemberManager.js:202]: Remove return or await

File: backend/lib/services/members/index.js [hunk 39]
```diff
@@ -32,3 +32,8 @@ exports.remove = async function ({ user, namespace, name }) {
   const memberManager = await MemberManager.create(user, namespace)
   return memberManager.delete(name)
 }
+
+exports.rotateSecret = async function ({ user, namespace, name }) {
+  const memberManager = await MemberManager.create(user, namespace)
+  memberManager.rotateServiceAccountSecret(name)
```
Comment (holgerkoser) [inline backend/lib/services/members/index.js:38]:
```suggestion
  await memberManager.rotateServiceAccountSecret(name)
```

File: backend/test/acceptance/api.members.spec.js [hunk 40]
```diff
@@ -393,5 +393,25 @@ describe('api', function () {
...
+      mockRequest.mockImplementationOnce(fixtures.projects.mocks.get())
+      mockRequest.mockImplementationOnce(fixtures.serviceaccounts.mocks.list())
+      mockRequest.mockImplementationOnce(fixtures.serviceaccounts.mocks.delete())
+
+      const res = await agent
+        .post(`/api/namespaces/${namespace}/members/${name}`)
+        .set('cookie', await user.cookie)
+        .send({
+          method: 'rotateSecret'
+        })
```
Comment (holgerkoser) [inline backend/test/acceptance/api.members.spec.js:409]:
```suggestion
        })
        .expect(200)
```

File: backend/test/acceptance/api.members.spec.js [hunk 41]
```diff
@@ -393,5 +393,25 @@ describe('api', function () {
 
       expect(res.body).toMatchSnapshot()
     })
+
+    it('should rotate a service account secret', async function () {
+      const name = 'system:serviceaccount:garden-foo:robot'
+
+      mockRequest.mockImplementationOnce(fixtures.projects.mocks.get())
+      mockRequest.mockImplementationOnce(fixtures.serviceaccounts.mocks.list())
+      mockRequest.mockImplementationOnce(fixtures.serviceaccounts.mocks.delete())
```
Comment (holgerkoser) [inline backend/test/acceptance/api.members.spec.js:402]:
```suggestion
      mockRequest.mockImplementationOnce(fixtures.secrets.mocks.delete())
```

File: backend/test/acceptance/api.members.spec.js [hunk 42]
```diff
@@ -393,5 +393,25 @@ describe('api', function () {
...
+        .post(`/api/namespaces/${namespace}/members/${name}`)
+        .set('cookie', await user.cookie)
+        .send({
+          method: 'rotateSecret'
+        })
+
+      expect(mockRequest).toBeCalledTimes(3)
+      expect(mockRequest.mock.calls).toMatchSnapshot()
+
+      expect(res.body).toMatchSnapshot()
```
Comment (holgerkoser) [inline backend/test/acceptance/api.members.spec.js:414]:
```suggestion
```

**Review by holgerkoser — APPROVED** [review]
/lgtm

**Review by holgerkoser — APPROVED** [review]
/lgtm

**Pull request merged**